raster/r.univar/r.univar_main.c

```diff
@@ -379,7 +408,215 @@ process_raster(univar_stat * stats, int fd, int fdz, const struct Cell_head *reg
+
+    const RASTER_MAP_TYPE map_type = Rast_get_map_type(fd[0]);
+    const size_t value_sz = Rast_cell_size(map_type);
+    void **raster_row;
+    CELL **zoneraster_row = NULL;
+    raster_row = G_malloc(threads * sizeof(void*));
+    if (n_zones) {
+        zoneraster_row = G_malloc(threads * sizeof(CELL*));
+    }
+    for (int i = 0; i < threads; i++) {
+        if (NULL == (raster_row[i] = Rast_allocate_buf(map_type))) {
+            return;
+        }
+        if (n_zones != 0 && (NULL == (zoneraster_row[i] = Rast_allocate_c_buf()))) {
+            return;
+        }
+    }
+
+    int n_alloc = n_zones ? n_zones : 1;
+    size_t **n = G_malloc(threads * sizeof(size_t*));
+    double **sum = G_malloc(threads * sizeof(double*));
+    double **sumsq = G_malloc(threads * sizeof(double*));
+    double **sum_abs = G_malloc(threads * sizeof(double*));
+    double **min = G_malloc(threads * sizeof(double*));
+    double **max = G_malloc(threads * sizeof(double*));
+    size_t **size = G_malloc(threads * sizeof(double*));
+
+    for (int i = 0; i < threads; i++) {
+        n[i] = G_calloc(n_alloc, sizeof(size_t));
+        sum[i] = G_calloc(n_alloc, sizeof(double));
+        sumsq[i] = G_calloc(n_alloc, sizeof(double));
+        sum_abs[i] = G_calloc(n_alloc, sizeof(double));
+        size[i] = G_calloc((n_alloc), sizeof(size_t));
+        min[i] = G_malloc((n_alloc) * sizeof(double));
+        max[i] = G_malloc((n_alloc) * sizeof(double));
+        for (int j = 0; j < n_alloc; j++) {
+            max[i][j] = DBL_MIN;
+            min[i][j] = DBL_MAX;
+        }
+    }
+
+    omp_lock_t *minmax = G_malloc(n_alloc * sizeof(omp_lock_t));
+    for (int i = 0; i < n_alloc; i++) {
+        omp_init_lock(&(minmax[i]));
+    }
+
+    unsigned int work = 0;
+    #pragma omp parallel default(shared)
+    {
+    int id = omp_get_thread_num();
+    unsigned int row;
+
+    #pragma omp for schedule(static)
+    for (row = 0; row < rows; row++) {
+	void *ptr;
+	CELL *zptr = NULL;
+	unsigned int col;
+
+	Rast_get_row(fd[id], raster_row[id], row, map_type);
+	if (n_zones) {
+	    Rast_get_c_row(fdz[id], zoneraster_row[id], row);
+	    zptr = zoneraster_row[id];
+	}
+
+	ptr = raster_row[id];
+
+    for (col = 0; col < cols; col++) {
+	    double val;
+	    int zone = 0;
+
+	    if (n_zones) {
+		/* skip NULL cells in zone map */
+		if (Rast_is_c_null_value(zptr)) {
+		    ptr = G_incr_void_ptr(ptr, value_sz);
+		    zptr++;
+		    continue;
+		}
+		zone = *zptr - zone_info.min;
+	    }
+
+	    /* count all including NULL cells in input map */
+	    /* stats[zone].size++; */
+        size[id][zone]++;
+	    
+	    /* can't do stats with NULL cells in input map */
+	    if (Rast_is_null_value(ptr, map_type)) {
+		ptr = G_incr_void_ptr(ptr, value_sz);
+		if (n_zones)
+		    zptr++;
+		continue;
+	    }
+
+	    val = ((map_type == DCELL_TYPE) ? *((DCELL *) ptr)
+			  : (map_type == FCELL_TYPE) ? *((FCELL *) ptr)
+			  : *((CELL *) ptr));
+
+	    sum[id][zone] += val;
```

Comment:
Would it be worth to go here with Kahan–Babuska algorithm?

Reply:
I think we can add this, but it is not the original behavior before this parallelization, perhaps we can do this in the next PR?